File: examples_pio/Wasm_Advanced/wasm_apps/assemblyscript/arduino.ts
export const LOW  = 0;
export const HIGH = 1;

export const INPUT        = 0x0;
export const OUTPUT       = 0x1;
export const INPUT_PULLUP = 0x2;

export declare function millis(): u32;
export declare function delay(ms: u32): void;
export declare function pinMode(pin: u32, mode: u32): void;
export declare function digitalWrite(pin: u32, value: u32): void;
export declare function getPinLED(): u32;

@external("getGreeting")
declare function _getGreeting(ptr: usize, maxlen: usize): void;

@external("print")
declare function _print(ptr: usize, len: usize): void;

export function print(str: string): void {
  const buffer = String.UTF8.encode(str);
  _print(changetype<usize>(buffer), buffer.byteLength);
}

export function println(str: string): void {
  print(str + '\n');
}

export function getGreeting(): string {
  const arr = new ArrayBuffer(64);
  _getGreeting(changetype<usize>(arr), 64);
  return String.UTF8.decode(arr, true);
}

File: examples_pio/Wasm_Advanced/wasm_apps/tinygo/arduino/arduino.go
package arduino

const (
    LOW  = 0
    HIGH = 1

    INPUT           =  0
    OUTPUT          =  1
    INPUT_PULLUP    =  2
)

//go:wasm-module arduino
//go:export millis
func Millis() uint

//go:wasm-module arduino
//go:export delay
func Delay(ms uint)

//go:wasm-module arduino
//go:export pinMode
func PinMode(pin, mode uint)

//go:wasm-module arduino
//go:export digitalWrite
func DigitalWrite(pin, value uint)

//go:wasm-module arduino
//go:export getPinLED
func GetPinLED() uint

//go:wasm-module arduino
//go:export print
func Print(s string)

//go:wasm-module arduino
//go:export getGreeting
func _getGreeting(buf *byte, maxlen uint)

func Println(s string) {
    print(s)
    print("\n")
}

func GetGreeting() string {
    var buf = make([]byte, 64)
    _getGreeting(&buf[0], 64)
    // Find '\0'
    n := -1
    for i, b := range buf {
        if b == 0 {
            break
        }
        n = i
    }
    return string(buf[:n+1])
}
